File: scripts/nbody.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>

void compute_forces(double* ax, double* ay, double* az, const double* x, const double* y,
                    const double* z, const double* m, int N, double G) {
    for (int i = 0; i < N; i++) {
        double axi = 0.0, ayi = 0.0, azi = 0.0;
        double xi = x[i], yi = y[i], zi = z[i];

        for (int j = 0; j < N; j++) {
            if (i != j) {
                double rx = xi - x[j];
                double ry = yi - y[j];
                double rz = zi - z[j];
                double d = sqrt(rx * rx + ry * ry + rz * rz);
                double c = G * m[j] / (d * d * d);
                axi -= c * rx;
                ayi -= c * ry;
                azi -= c * rz;
            }
        }

        ax[i] = axi;
        ay[i] = ayi;
        az[i] = azi;
    }
}

int main() {
    // Test case: 4 particles at corners of a unit square
    double x4[] = {0.0, 1.0, 0.0, 1.0};
    double y4[] = {0.0, 0.0, 1.0, 1.0};
    double z4[] = {0.0, 0.0, 0.0, 0.0};
    double m4[] = {1.0, 1.0, 1.0, 1.0};
    double ax4[4], ay4[4], az4[4];

    compute_forces(ax4, ay4, az4, x4, y4, z4, m4, 4, 1.0);
    double checksum = 0.0;
    for (int i = 0; i < 4; i++)
        checksum += sqrt(ax4[i] * ax4[i] + ay4[i] * ay4[i] + az4[i] * az4[i]);
    printf("Checksum: %g\n", checksum);

    // Benchmark
    int N = 10000;
    double* x = malloc(N * sizeof(double));
    double* y = malloc(N * sizeof(double));
    double* z = malloc(N * sizeof(double));
    double* m = malloc(N * sizeof(double));
    double* ax = malloc(N * sizeof(double));
    double* ay = malloc(N * sizeof(double));
    double* az = malloc(N * sizeof(double));

    srand(42);
    for (int i = 0; i < N; i++) {
        x[i] = (double)rand() / RAND_MAX;
        y[i] = (double)rand() / RAND_MAX;
        z[i] = (double)rand() / RAND_MAX;
        m[i] = 1.0;
    }

    struct timeval start, end;
    gettimeofday(&start, NULL);
    compute_forces(ax, ay, az, x, y, z, m, N, 1.0);
    gettimeofday(&end, NULL);

    double elapsed = (end.tv_sec - start.tv_sec) + (end.tv_usec - start.tv_usec) / 1e6;
    printf("Time: %.4f s\n", elapsed);

    // Checksum to prevent compiler from optimizing away the computation
    double bench_checksum = 0.0;
    for (int i = 0; i < N; i++)
        bench_checksum += ax[i] + ay[i] + az[i];
    printf("Bench checksum: %g\n", bench_checksum);

    free(x);
    free(y);
    free(z);
    free(m);
    free(ax);
    free(ay);
    free(az);
    return 0;
}

File: scripts/nbody.py
import math
import random
import time


def compute_forces(ax, ay, az, x, y, z, m, G=1.0):
    N = len(m)
    for i in range(N):
        axi, ayi, azi = 0.0, 0.0, 0.0
        xi, yi, zi = x[i], y[i], z[i]

        for j in range(N):
            if i != j:
                rx = xi - x[j]
                ry = yi - y[j]
                rz = zi - z[j]
                d = math.sqrt(rx * rx + ry * ry + rz * rz)
                c = G * m[j] / (d * d * d)
                axi -= c * rx
                ayi -= c * ry
                azi -= c * rz

        ax[i] = axi
        ay[i] = ayi
        az[i] = azi


# Test case: 4 particles at corners of a unit square
x4 = [0.0, 1.0, 0.0, 1.0]
y4 = [0.0, 0.0, 1.0, 1.0]
z4 = [0.0, 0.0, 0.0, 0.0]
m4 = [1.0, 1.0, 1.0, 1.0]
ax4, ay4, az4 = [0.0] * 4, [0.0] * 4, [0.0] * 4

compute_forces(ax4, ay4, az4, x4, y4, z4, m4)
checksum = sum(math.sqrt(ax4[i] ** 2 + ay4[i] **
               2 + az4[i] ** 2) for i in range(4))
print(f"Checksum: {checksum}")

# Benchmark
N = 10_000
random.seed(42)
x = [random.random() for _ in range(N)]
y = [random.random() for _ in range(N)]
z = [random.random() for _ in range(N)]
m = [1.0] * N
ax, ay, az = [0.0] * N, [0.0] * N, [0.0] * N

start = time.time()
compute_forces(ax, ay, az, x, y, z, m)
elapsed = time.time() - start
print(f"Time: {elapsed:.4f} s")

# Checksum to prevent compiler from optimizing away the computation
bench_checksum = sum(ax) + sum(ay) + sum(az)
print(f"Bench checksum: {bench_checksum}")
